### PR TITLE
Replaced MPI communications and updated FEMNIST example

### PR DESCRIPTION
**Pull request by yim0331**
- The collective communication of MPI for the NN parameters from multiple clients exceeded the maximum message size allowed in MPI. We replaced the collective communication between the server and the clients with the point-to-point communication.
- FEMINIST Example updates include the issues [#112](https://github.com/APPFL/APPFL/issues/112) and [#109](https://github.com/APPFL/APPFL/pull/109)

### CODE REVIEW

File: examples/models/resnet.py [hunk 1]
```diff
@@ -1,7 +1,128 @@
...
+
+    
     if num_classes is not None:
-        return models.resnet18(pretrained=False, num_classes=num_classes)
+        resnetmodel = models.resnet18(pretrained=False, num_classes=num_classes)
     else:
-        return models.resnet18(pretrained=True)
+        resnetmodel = models.resnet18(pretrained=True)
+
+    if num_channel == 1:
```
Comment (kibaekkim) [inline examples/models/resnet.py:115]: Is this the case for FEMNIST?

File: examples/models/resnet.py [hunk 2]
```diff
@@ -1,7 +1,128 @@
...
+
+    if num_channel == 1:        
+        
+        model = ResNet(block=BasicBlock, layers=[2, 2, 2, 2],num_classes=num_classes, grayscale=True)
+        
+        # model = nn.Sequential()
+        # model.append(nn.ZeroPad2d(244-num_pixel)) #deafult input size for resnet18 is 244
+        # resnetmodel.conv1 = nn.Conv2d(1, 64, kernel_size=7, stride=2, padding=3,
+        #                        bias=False)
+        # model.append(resnetmodel)
```
Comment (kibaekkim) [inline examples/models/resnet.py:123]: If we don't use these lines later, please remove these.

File: examples/models/resnet.py [hunk 3]
```diff
@@ -1,7 +1,128 @@
...
     if num_classes is not None:
-        return models.resnet18(pretrained=False, num_classes=num_classes)
+        resnetmodel = models.resnet18(pretrained=False, num_classes=num_classes)
     else:
-        return models.resnet18(pretrained=True)
+        resnetmodel = models.resnet18(pretrained=True)
+
+    if num_channel == 1:        
+        
+        model = ResNet(block=BasicBlock, layers=[2, 2, 2, 2],num_classes=num_classes, grayscale=True)
```
Comment (yim0331) [inline examples/models/resnet.py:114]: This is the Resnet Model for FEMNIST

File: src/appfl/run_mpi.py [hunk 4]
```diff
@@ -266,11 +278,9 @@ def run_client(
...
             client.model.load_state_dict(global_state)
 
-            ## client update
-            local_states[cid] = client.update()
-
-        """ Send "local_states" to a server """
-        comm.gather(local_states, root=0)
+            ## client update     
+            ls = client.update()                            
+            req = comm.send(ls, 0, tag=cid)
```
Comment (kibaekkim) [inline src/appfl/run_mpi.py:283]: I believe we can use non-blocking communication here. Please use `isend`.

**Review by kibaekkim — CHANGES_REQUESTED** [review]
@yim0331 Please add the description for class (e.g., `ResNet18_Weights`) and functions and in-line comments to the files in `examples`. In particular, the examples should provide a good guidance to users for how to use the package.

The modification to the `algorithm` part needs @minseok-ryu 's review.

**Comment by kibaekkim** [comment]
@yim0331 The PR should be tested with the new `build` CI. Please merge `main` to this branch.

**Review by minseok-ryu — APPROVED** [review]
Minor modifications in Algorithm look good to me.

**Review by kibaekkim — APPROVED** [review]

**Pull request merged**